File: customer-service/src/main/java/com/wks/servicemarketplace/customerservice/config/GraphQLDataFetcherExceptionHandler.java
package com.wks.servicemarketplace.customerservice.config;

import com.wks.servicemarketplace.customerservice.adapters.graphql.GraphQLUseCaseError;
import com.wks.servicemarketplace.customerservice.adapters.graphql.SanitizedError;
import com.wks.servicemarketplace.customerservice.core.exceptions.CoreException;
import com.wks.servicemarketplace.customerservice.core.exceptions.ErrorType;
import com.wks.servicemarketplace.customerservice.core.exceptions.ValidationException;
import graphql.ExceptionWhileDataFetching;
import graphql.GraphQLError;
import graphql.execution.DataFetcherExceptionHandler;
import graphql.execution.DataFetcherExceptionHandlerParameters;
import graphql.execution.DataFetcherExceptionHandlerResult;

public class GraphQLDataFetcherExceptionHandler implements DataFetcherExceptionHandler {

    @Override
    public DataFetcherExceptionHandlerResult onException(DataFetcherExceptionHandlerParameters handlerParameters) {
        final ExceptionWhileDataFetching fetchError = new ExceptionWhileDataFetching(handlerParameters.getPath(), handlerParameters.getException(), handlerParameters.getSourceLocation());
        final GraphQLError error = getGraphQLError(fetchError);
        return DataFetcherExceptionHandlerResult.newResult(error).build();
    }

    private GraphQLError getGraphQLError(ExceptionWhileDataFetching fetchError) {
        if (fetchError.getException() instanceof CoreException) {
            return new GraphQLUseCaseError((CoreException) fetchError.getException(), fetchError);
        }
        if (fetchError.getException().getCause() instanceof CoreException) {
            return new GraphQLUseCaseError((CoreException) fetchError.getException().getCause(), fetchError);
        }
        if (fetchError.getException() instanceof ValidationException) {
            final ValidationException validationException = (ValidationException) fetchError.getException();
            return new GraphQLUseCaseError(new CoreException(
                    ErrorType.VALIDATION,
                    validationException.getMessage(),
                    validationException.getFields(),
                    validationException
            ), fetchError);
        }
        return new SanitizedError(fetchError);
    }
}

File: service-provider-service/src/main/kotlin/com/wks/servicemarketplace/serviceproviderservice/adapters/graphql/GraphQLDataFetcherExceptionHandler.kt
package com.wks.servicemarketplace.serviceproviderservice.adapters.graphql

import com.wks.servicemarketplace.serviceproviderservice.core.exceptions.CoreException
import graphql.ExceptionWhileDataFetching
import graphql.execution.DataFetcherExceptionHandler
import graphql.execution.DataFetcherExceptionHandlerParameters
import graphql.execution.DataFetcherExceptionHandlerResult

class GraphQLDataFetcherExceptionHandler : DataFetcherExceptionHandler {
    override fun onException(handlerParameters: DataFetcherExceptionHandlerParameters): DataFetcherExceptionHandlerResult {
        return handlerParameters.let {
            ExceptionWhileDataFetching(it.path, it.exception, it.sourceLocation)
        }.let {
            when (it) {
                is CoreException -> (it.exception as CoreException).toDataFetcherExceptionHandlerResult()
                else -> DataFetcherExceptionHandlerResult.newResult(GraphQLSanitizedException(it)).build()
            }
        }
    }
}

fun CoreException.toDataFetcherExceptionHandlerResult(): DataFetcherExceptionHandlerResult {
    return DataFetcherExceptionHandlerResult.newResult(
            GraphQLCoreException(this.message, this.fields, this.errorType)
    ).build()
}